src/emqx_router.erl

```diff
@@ -248,46 +281,38 @@ delete_trie_route(Route = #route{topic = Topic}) ->
     end.
 
 %% @private
--spec(maybe_trans(function(), list(any())) -> ok | {error, term()}).
-maybe_trans(Fun, Args) ->
-    case persistent_term:get(emqx_route_lock_type) of
+-spec(maybe_trans(pid(), function(), list(any())) -> ok | {error, term()}).
+maybe_trans(Worker, Fun, Args) ->
+    case persistent_term:get(emqx_route_lock_type, key) of
         key ->
-            trans(Fun, Args);
+            trans(Worker, Fun, Args);
         global ->
             lock_router(),
             try mnesia:sync_dirty(Fun, Args)
             after
                 unlock_router()
             end;
         tab ->
-            trans(fun() ->
+            trans(Worker,
+                  fun() ->
                           emqx_trie:lock_tables(),
                           apply(Fun, Args)
                   end, [])
     end.
 
--spec(trans(function(), list(any())) -> ok | {error, term()}).
-trans(Fun, Args) ->
+-spec(trans(pid(), function(), list(any())) -> ok | {error, term()}).
+trans(Worker, Fun, Args) when is_pid(Worker) ->
     %% trigger selective receive optimization of compiler,
     %% ideal for handling bursty traffic.
-    Ref = erlang:make_ref(),
-    Owner = self(),
-    {WPid, RefMon} = spawn_monitor(
-                       fun() ->
-                               Res = case mnesia:transaction(Fun, Args) of
-                                         {atomic, Ok} -> Ok;
-                                         {aborted, Reason} -> {error, Reason}
-                                     end,
-                               Owner ! {Ref, Res}
-                       end),
+    MonRef = erlang:monitor(process, Worker),
+    Worker ! {work, MonRef, Fun, Args},
     receive
-        {Ref, TransRes} ->
-            receive
-                {'DOWN', RefMon, process, WPid, normal} -> ok
-            end,
+        {MonRef, TransRes} ->
+            %% note! Caller should remember to flush 'DownMsg'
+            true = erlang:demonitor(MonRef),
```

Comment:
if we do not flush here, the DOWN message will be leaked to the caller, result in a error log.
```
handle_info(Info, State) ->
    ?LOG(error, "Unexpected info: ~p", [Info]),
    {noreply, State}.
```

Reply:
Ok. I think I should flush the message in place here since gen.erl do the same.

Reply:
but,,, the idea of spawning long-lived process is for us NOT having to monitor + flush-demonitor here.
so the receive is optimised by the compiler.

Reply:
I have checked that the receive is optimized by the compiler.

Reply:
sorry, I forget what I have investigated last week. as notes in the code says, it is not good idea to flush here.

it is ok for gen.erl to flush since gen_server is mostly a long-living process and flush only happens when demon is failed (return false.)

Reply:
I just spend some time investigating,  erlang:demonitor/2 is smart enough to skip the flush if the demonitor call (C code) success (returns true) even if the `flush` flag is set in the options.

In our case,  with long-living process, demonitor most likely returns true so no flush or extra 'receive' call is involved.